### PR TITLE
add checker for return in finally

### LINKED ISSUE #8260: Prevent usage of `return`, `yield` in `try/finally`
### Current problem

Code in `finally` blocks will always be executed. If there's a `return` inside the `try`, it's possible to overwrite this return value with a `return` in the `finally`.

```py
def add(a: int, b: int) -> int:
    try:
        return a + b
    finally:
        return 0


if __name__ == "__main__":
    print(add(1, 2))  # those args will be ignored; "0" is printed
```

### Desired solution

There should not be any `return` or `yield` statements in a finally block.

### Additional context

https://docs.python.org/3/reference/compound_stmts.html#finally-clause

> The return value of a function is determined by the last return statement executed.

### PR DESCRIPTION
**Pull request by hofrob**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #8260

### CODE REVIEW

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/refactoring/refactoring_checker.py:689]: It's possible to use ``for child_node in node.nodes_of_class(nodes.Yield, nodes.Return)`` to recover only the yield/return nodes in a generator.

**Review by Pierre-Sassoulas — DISMISSED** [review]
Thank you for working on this, it looks pretty good already !

**Comment by hofrob** [comment]
I'm not too happy with those examples. The problem is, that pylint should probably not make recommendations on how to fix it. The only real suggestion this checker could offer, would be "the `finally` should only contain shutdown/close commands". If there's a good source where it says what to do in more detail, I'd add it.

I added this checker to the `pylint/checkers/refactoring/refactoring_checker.py` module, but if you have another suggestion, I'll gladly move it. I'm also walking through all statements in a `finalbody`. This might impact performance, though I hope that projects do not have a lot of logic inside of their `finally` clauses :grimacing:.

Other than that, let me know if I should improve the wording or anything else really.

The fact that this new checker only produces one more message in all the checked repos is a good sign in my opinion. I'm not sure if the usage of yield in finally in pycopg2 is a legitimate one?

**Comment by Pierre-Sassoulas** [comment]
It's a very good sign when there's only one new message raised and it's hard to tell if it's a false positive :)

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (hofrob) [reply pylint/checkers/refactoring/refactoring_checker.py]: Oh my, that is good to know. Thanks!

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (hofrob) [reply pylint/checkers/refactoring/refactoring_checker.py]: I moved it to `pylint/checkers/base/basic_checker.py`

**Comment by hofrob** [comment]
After looking at the generated docs, I thought adding code to the bad example might be a bit more helpful.

File: doc/data/messages/r/return-or-yield-in-finally/bad.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/r/return-or-yield-in-finally/bad.py:1]: It's possible to create multiple files in a directory too, for multiple examples (a bad directory and a good directory) I think this would be better. We've never done it before so we'll have to deal with the surprises if they are any :)

Comment (hofrob) [reply doc/data/messages/r/return-or-yield-in-finally/bad.py]: Ok, let's see what happens :grinning:

Comment (hofrob) [reply doc/data/messages/r/return-or-yield-in-finally/bad.py]: https://pylint--8440.org.readthedocs.build/en/8440/user_guide/messages/warning/return-or-yield-in-finally.html

:champagne:

File: doc/data/messages/r/return-or-yield-in-finally/good/return.py
Comment (Pierre-Sassoulas) [inline doc/data/messages/r/return-or-yield-in-finally/good/return.py:2]:
```suggestion
    fruits = ["kiwi", "pineapple"]
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
I like the embedded comment in the doc

**Comment by nickdrozd** [comment]
The primer is showing one hit that looks like a false positive:

```python
    def _exit_gen(self) -> PQGen[None]:
        """
        Exit current pipeline by sending a Sync and fetch back all remaining results.
        """
        try:
            self._enqueue_sync()
            yield from self._communicate_gen()
        finally:
            # No need to force flush since we emitted a sync just before.
            yield from self._fetch_gen(flush=False)
```

That function yields at least one item in any case, and possibly more. But I don't see any bug there.

I think the warning should only trigger if there is a `return` somewhere.

**Comment by hofrob** [comment]
> ```python
>     def _exit_gen(self) -> PQGen[None]:
>         """
>         Exit current pipeline by sending a Sync and fetch back all remaining results.
>         """
>         try:
>             self._enqueue_sync()
>             yield from self._communicate_gen()
>         finally:
>             # No need to force flush since we emitted a sync just before.
>             yield from self._fetch_gen(flush=False)
> ```

I'm trying to understand why this `yield` is in a `finally` clause. It more or less looks like they're trying to skip "bare-except" lint warnings this way 😀. Because that's what this actually does, right?

So I'm still not convinced that this is actually a false positive. `yield from self._fetch_gen(flush=False)` should come after the `try`/`except` with actual exceptions they want to catch.

😕

**Comment by nickdrozd** [comment]
> I'm trying to understand why this `yield` is in a `finally` clause. It more or less looks like they're trying to skip "bare-except" lint warnings this way grinning. Because that's what this actually does, right?

I'm not sure why it's like that. It's not skipping bare-except because it doesn't actually catch anything. So if there's an exception in `try`, it will interfere with the `yield` in `finally`? But `self._fetch_gen` needs to get called in any case, which might result in side effects.

Barring side effects, I can't think of a good reason to do it this way. So maybe it isn't a false positive. But I could be wrong.

File: pylint/checkers/base/basic_checker.py
Comment (jacobtylerwalls) [inline pylint/checkers/base/basic_checker.py:273]: Speaking a bit out of my depth here, but I think if `return`ing from a generator, the return value won't be overwritten strictly speaking, it will just be passed to a coroutine. So this is slightly incorrectly described for `yield`s, leaving me to suggest to back out all the yield-related stuff.

File: doc/data/messages/r/return-or-yield-in-finally/bad/yield.py
Comment (jacobtylerwalls) [inline doc/data/messages/r/return-or-yield-in-finally/bad/yield.py:11]: Strikes me as a little contrived, who's to say that interleaving None between those fruits is not intended/useful? It's not a crazy use case to need to interleave a constant into a every other position in a sequence.

File: doc/data/messages/r/return-or-yield-in-finally/bad/return.py
Comment (jacobtylerwalls) [inline doc/data/messages/r/return-or-yield-in-finally/bad/return.py:7]: Great!

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
Thanks for this!

Agree with @nickdrozd we should just deal with `return`. I don't think `yield` is really a problem in `finally`. I'd be happy to merge this if just affecting `return`.

**Comment by hofrob** [comment]
Ok, I see that including `yield` is a step too far. I removed references to `yield` now but will do a _final_ (:wink:) check and wait for the CI to finish.

Thanks for chiming in! This PR has bothered me for weeks. But I totally agree now, that this should only look at `return` statements.

File: tests/functional/r/return_in_finally.py
Comment (jacobtylerwalls) [inline tests/functional/r/return_in_finally.py:32]: What about an example with `return` in the finally block but lacks a `return` in the try block. (And would you consider that something that should still emit a message?)

Comment (jacobtylerwalls) [reply tests/functional/r/return_in_finally.py:32]: I'm imagining a legit use case like:

```python
def divide(numerator, denominator):
    try:
        result = numerator / denominator
    except ZeroDivisionError:
        try:
            send_error()
        finally:
            return
    return result
```

Comment (hofrob) [reply tests/functional/r/return_in_finally.py:32]: A `return` in a `finally` will _always_ overwrite the return value. It doesn't matter if there's a `return` inside the `try` itself.

So even if your function catches an error and continues without raising any other exceptions, it will execute the `finally` and if there was a return statement, it will return this value and discard whatever was returned afterwards.

```py
def foo():
    try:
        raise Exception
    finally:
        return 1

    return 2

print(foo())
```

This prints `1`.

Comment (hofrob) [reply tests/functional/r/return_in_finally.py:32]: Wait no. Scrap that (ignore my example).

Let me think about it for a bit.

Comment (jacobtylerwalls) [reply tests/functional/r/return_in_finally.py:32]: But in my example, only if the inner `try` executes, right?
```python
>>> divide(1, 4)
0.25
```

I think I'm asking what you think of checking whether or not the try/finally in question is guarded under `except`.

Comment (hofrob) [reply tests/functional/r/return_in_finally.py:32]: Do you have any real world examples? If I see something like this in a MR/PR, I'd question if there's something wrong in general here. For example, the `send_error()` method should probably take care of any errors. So it would look like this:

```py
def divide(numerator, denominator):
    result = None
    try:
        result = numerator / denominator
    except ZeroDivisionError:
        send_error()  # send_error should handle its own exceptions
    return result
```

So I can't really think of a situation where this would make sense to have a `return` in a `finally`, let alone a `try` in an `except` clause.

**Comment by Pierre-Sassoulas** [comment]
The doc fail is due to a bug that was fixed on main :)

File: tests/functional/r/return_in_finally.py
Comment (jacobtylerwalls) [reply tests/functional/r/return_in_finally.py:32]: I think the most realistic case to consider would be when you're not in control of `send_error()`,  or if it's more idiomatic for send_error to raise documented, idiomatic exceptions like NotAuthorized or Http500Exception or whatnot. But I don't think we need to block a merge over this; a user in this scenario can just disable `return-in-finally` since they are doing something a little brittle.

File: pylint/checkers/base/basic_checker.py
Comment (jacobtylerwalls) [inline pylint/checkers/base/basic_checker.py:778]: We add confidence to new messages
```suggestion
                self.add_message("return-in-finally", node=return_node, confidence=HIGH)
```

File: tests/functional/r/return_in_finally.txt
Comment (jacobtylerwalls) [inline tests/functional/r/return_in_finally.txt:4]:
```suggestion
return-in-finally:8:8:8:24:second_favorite:'return' shadowed by the 'finally' clause.:HIGH
return-in-finally:18:12:18:18:all_favorites:'return' shadowed by the 'finally' clause.:HIGH
return-in-finally:29:16:29:40:more_favorites:'return' shadowed by the 'finally' clause.:HIGH
return-in-finally:41:20:41:56:even_more_favorites:'return' shadowed by the 'finally' clause.:HIGH
```

**Review by jacobtylerwalls — APPROVED** [review]

File: tests/functional/r/return_in_finally.py
Comment (hofrob) [reply tests/functional/r/return_in_finally.py:32]: I know we're drifting off-topic now, but in such a case I would just have a module that takes care of external stuff I can't handle and call my "proxy"-methods instead.

I know it's not a thorough explanation of my thought process though and there still might be edge cases where one has to live with such a situation.

File: pylint/checkers/base/basic_checker.py
Comment (hofrob) [reply pylint/checkers/base/basic_checker.py]: Thanks! Done.

**Review by jacobtylerwalls — APPROVED** [review]
Thanks for the contribution!

**Comment by Pierre-Sassoulas** [comment]
Great new check, thank you !

**Pull request merged**
